### PR TITLE
Update file change history before calling MappableCommand

### LINKED ISSUE #4719: file-name element doesn't update when saving under INS mode
### Summary

Saving file while in insert mode doesn't properly update the file-name `[+]`. It seems to mostly be just a graphical error, but on Linux (at least the machine I was SSHed into), saving in insert mode saves the file, but helix doesn't let you quit.
(And the saving while in insert mode is done with key mapping)

### Reproduction Steps

 
`config.toml`:
```
[keys.insert]
 C-s = ":w"
```
I tried this:

1. Open file `hx temp.txt`
2. Enter ins mode `i`
3. Enter text "hello world"
4. While still in insert mode try to save `C-s`
5. Quit out of file `ESC`, `:q`

I expected this to happen:
File would save and [+] would go away

Instead, this happened:
On windows:
File successfully saved, but the `[+]` did not successfully update on the file-name status bar.

On Linux (at least though SSH)
`[+]` graphic did not update, but the following message appeared under the status bar: `'temp.txt' written 4L 34B`. 
After exiting INS mode and trying to `:q` I get the following message: `1 unsaved buffer(s) remaining: ["temp.txt"]`. Force quitting exits the editor and the file seems to have saved correctly.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2022-11-12T04:17:39.287 helix_view::clipboard::provider [WARN] No clipboard provider found! Yanking and pasting will be internal to Helix
2022-11-12T04:17:53.244 helix_view::editor [ERROR] editor error: 1 unsaved buffer(s) remaining: ["temp.txt"]
2022-11-12T04:18:09.911 helix_view::clipboard::provider [WARN] No clipboard provider found! Yanking and pasting will be internal to Helix
2022-11-12T04:20:13.897 helix_view::clipboard::provider [WARN] No clipboard provider found! Yanking and pasting will be internal to Helix
2022-11-12T04:20:23.360 helix_view::editor [ERROR] editor error: 1 unsaved buffer(s) remaining: ["temp.txt"]
2022-11-12T04:20:28.330 helix_view::clipboard::provider [WARN] No clipboard provider found! Yanking and pasting will be internal to Helix
2022-11-12T04:20:41.857 helix_view::editor [ERROR] editor error: 1 unsaved buffer(s) remaining: ["temp.txt"]
2022-11-12T04:24:58.337 helix_view::clipboard::provider [WARN] No clipboard provider found! Yanking and pasting will be internal to Helix
2022-11-12T04:25:09.863 helix_view::editor [ERROR] editor error: 1 unsaved buffer(s) remaining: ["temp.txt"]
2022-11-12T04:26:34.739 helix_view::clipboard::provider [WARN] No clipboard provider found! Yanking and pasting will be internal to Helix
2022-11-12T04:28:19.549 helix_view::editor [ERROR] editor error: 1 unsaved buffer(s) remaining: [".cache/helix/helix.log"]
```

</details>

### Platform

Windows + Linux over SSH

### Terminal Emulator

Windows Terminal & xterm-256color over SSH

### Helix Version

helix 22.08.1 (001e4e30)

### PR DESCRIPTION
**Pull request by swork1**
Before we run a mappable command we should update the current changes to history.

 This fixes #4719

### CODE REVIEW

**Comment by the-mikedavis** [comment]
Alternatively, keybinds that call typable commands from insert-mode could include `commit_undo_checkpoint` before the call to save

**Comment by the-mikedavis** [comment]
I meant that it could be done in keymaps:

```toml
[keys.insert]
C-s = ["commit_undo_checkpoint", ":w"]
```

but I suppose there isn't really a downside to doing this for all typable commands. Mostly this will just be a no-op since you need to enter normal/select modes to execute typable commands and that also saves a checkpoint

**Comment by archseer** [comment]
This breaks how the undo is supposed to work though, we've had this issue & PR before: https://github.com/helix-editor/helix/pull/2883#pullrequestreview-1029452126 https://github.com/helix-editor/helix/issues/3501

**Comment by archseer** [comment]
I think this needs some discussion but maybe having the undo state broken into two undo steps is preferable to having this broken.

File: helix-term/src/commands.rs [hunk 1]
```diff
@@ -167,6 +167,7 @@ impl MappableCommand {
             Self::Typable { name, args, doc: _ } => {
                 let args: Vec<Cow<str>> = args.iter().map(Cow::from).collect();
                 if let Some(command) = typed::TYPABLE_COMMAND_MAP.get(name.as_str()) {
+                    commit_undo_checkpoint(cx);
```
Comment (archseer) [inline helix-term/src/commands.rs:170]: Should only be called if in insert mode, together with a comment explaining why this is needed in the first place

**Comment by pascalkuthe** [comment]
I don't think this is the right solution to the problem. Some commands are fine to call from insertmode with creating a checkpoint. These commands should not be changed to create a checkpoint. 

closing this one out as stale, thank you for contributing!

**Pull request closed without merging**